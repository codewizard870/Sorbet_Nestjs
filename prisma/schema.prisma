// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model User {
  id                    String       @id @default(auto()) @map("_id") @db.ObjectId
  nearWallet            String?      @unique
  firstName             String?
  lastName              String?
  email                 String       @unique
  bio                   String?
  confirmationCode      String?      @unique
  profileImage          String?
  connections           String[]     @db.ObjectId @default([])
  connection_requests   String[]     @db.ObjectId @default([])
  followers             String[]     @db.ObjectId @default([])
  following             String[]     @db.ObjectId @default([])
  jobProfile            JobProfile[]
  token                 Token?
  location              Location[]
  post                  Post[]
  groupIDs              String[]     @db.ObjectId
  groups                Group[]      @relation(name: "members", fields: [groupIDs], references: [id])
  createdAt             DateTime     @default(now())
}

model JobProfile {
  id        String        @id @default(auto()) @map("_id") @db.ObjectId
  name      String?
  type      String?
  createdAt DateTime      @default(now())
  //relation with user
  user      User?         @relation(fields: [userId], references: [id])
  userId    String?       @db.ObjectId
}

enum Status {
  Pending  @map("pending")
  Active   @map("active")
  Disabled @map("disabled")

  @@map("Status")
}

model SocialMedia {
  id          String      @id @default(auto()) @map("_id") @db.ObjectId
  socialMedia String
  link        String
}

model Token {
  id        String        @id @default(auto()) @map("_id") @db.ObjectId
  token     String
  createdAt DateTime      @default(now())

  //relation with user
  user   User             @relation(fields: [userId], references: [id])
  userId String           @unique @db.ObjectId
}

model Event {
  id          String      @id @default(auto()) @map("_id") @db.ObjectId
  event_image String
  event_link  String
  name String
  description String
  timezone    DateTime
  start_date  DateTime
  end_date    DateTime?
  location    Location[]
  //relation with post
  post        Post        @relation(fields: [postId], references: [id], onDelete: Cascade)
  postId      String      @db.ObjectId
}

model Gig {
  id            String    @id @default(auto()) @map("_id") @db.ObjectId
  timezone      DateTime
  start_date    DateTime
  end_date      DateTime?
  title         String
  description   String
  gig_price_min Int
  gig_price_max Int
  tags          String[]
  location      Location[]
  //relation with post
  post          Post      @relation(fields: [postId], references: [id], onDelete: Cascade)
  postId        String    @db.ObjectId
}

model Post {
  id        String        @id @default(auto()) @map("_id") @db.ObjectId
  timestamp DateTime
  text      String?
  content   Content
  gig       Gig[]
  event     Event[]
  blob      Blob[]
  location  Location[]
  //relation with user
  user      User          @relation(fields: [userId], references: [id], onDelete: Cascade)
  userId    String        @db.ObjectId
}

model Blob {
  id       String         @id @default(auto()) @map("_id") @db.ObjectId
  imageUrl String
  videoUrl String

  //relation with post
  post   Post             @relation(fields: [postId], references: [id], onDelete: Cascade)
  postId String           @unique @db.ObjectId
}

model Location {
  id            String    @id @default(auto()) @map("_id") @db.ObjectId
  location_type LocationType
  country       String
  province      String
  district      String
  city          String
  Latitude      Float
  Langitude     Float

// relation with post
  post   Post?            @relation(fields: [postId], references: [id], onDelete: Cascade)
  postId String?          @db.ObjectId

  // relation with user
  user         User?      @relation(fields: [userId], references: [id])
  userId       String?    @db.ObjectId

  // relation with event
  event         Event?    @relation(fields: [eventId], references: [id])
  eventId       String?   @db.ObjectId

  // relation with gig
  gig           Gig?      @relation(fields: [gigId], references: [id])
  gigId         String?   @db.ObjectId

  // relation with group
  group         Group?    @relation(fields: [groupId], references: [id])
  groupId       String?   @db.ObjectId

  // relation with collab
  // collab        Collab?   @relation(fields: [collabId], references: [id])
  // collabId      String?   @db.ObjectId
}

enum Content {
  Gig   @map("gig")
  Event @map("event")
  Blob  @map("blob")

  @@map("Content")
}

enum LocationType {
  Remote @map("remote")
  OnSite @map("onSite")

  @@map("LocationType")
}

model Contact {
  id               String   @id @default(auto()) @map("_id") @db.ObjectId
  userId           String
  contacted_userId String
  chat             Chat[]
}

model Chat {
  id        String          @id @default(auto()) @map("_id") @db.ObjectId
  message   String
  creatorId String
  createdAt DateTime        @default(now())
  updatedAt DateTime?       @default(now())

  //relation with contact
  contact   Contact         @relation(fields: [contactId], references: [id], onDelete: Cascade)
  contactId String          @db.ObjectId
}

model Collab {
  id               String   @id @default(auto()) @map("_id") @db.ObjectId
  collabId         String
  userId           String
  wallet_address   String
  public_key       String
  // location         Location[]
  createdAt        DateTime
  updatedAt        DateTime?
}

model Group {
  id               String   @id @default(auto()) @map("_id") @db.ObjectId
  name             String
  description      String?
  image            String?
  group_owner      String   @db.ObjectId
  userIDs          String[] @db.ObjectId
  members          User[]   @relation(name: "members", fields: [userIDs], references: [id])
  location         Location[]
  createdAt        DateTime @default(now())
  updatedAt        DateTime?
}
