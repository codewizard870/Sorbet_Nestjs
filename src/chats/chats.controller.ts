import {
  Controller,
  Request,
  Get,
  Post,
  Body,
  Patch,
  Param,
  Delete,
} from "@nestjs/common";
import { ApiBearerAuth, ApiTags } from "@nestjs/swagger";
import { ChatsService } from "./chats.service";
import { CreateChatDto } from "./dto/create-chat.dto";
import { UpdateChatDto } from "./dto/update-chat.dto";

@ApiBearerAuth()
@ApiTags("Chats")
@Controller("/swagger/chats")
export class ChatsController {
  constructor(private readonly chatsService: ChatsService) {}

  @Post()
  async create(@Body() createChatDto: CreateChatDto, @Request() req) {
    return await this.chatsService.create(createChatDto, req.user.id);
  }

  @Get()
  async findAll() {
    return await this.chatsService.getAll();
  }

  @Get(":id")
  async findOne(@Param("id") id: string) {
    return await this.chatsService.findOne(id);
  }

  @Get("findBycontactId/:contactId")
  async findByContactId(@Param("contactId") contactId: string) {
    return await this.chatsService.getChatByContactId(contactId);
  }

  @Get("findByuserId/:userId")
  async findByUserId(@Param("userId") userId: string) {
    return await this.chatsService.getChatByUserId(userId);
  }

  @Patch(":id")
  update(@Param("id") id: string, @Body() updateChatsDto: UpdateChatDto) {
    return this.chatsService.update(id, updateChatsDto);
  }

  @Delete(":id")
  async remove(@Param("id") id: string) {
    return await this.chatsService.remove(id);
  }

  @Get("searchMessages")
  async searchMessages(@Body() text: string) {
    return await this.chatsService.searchMessages(text)
  }
}
