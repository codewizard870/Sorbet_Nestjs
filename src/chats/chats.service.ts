import { BadRequestException, Injectable } from "@nestjs/common";
import { PrismaService } from "src/utils/prisma/prisma.service";
import { CreateChatDto } from "./dto/create-chat.dto";
import { UpdateChatDto } from "./dto/update-chat.dto";

@Injectable()
export class ChatsService {
  constructor(private prismaService: PrismaService) {}

  async create(data: any, userId: string) {
    try {
      const result = await this.prismaService.chat.create({
        data: {
          message: data.message,
          creatorId: userId,
          contactId: data.contactId,
        },
      })
      if (result) {
        return result
      } 
      else {
        throw new BadRequestException("Result not found")
      }
    } 
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async getAll() {
    try {
      const allChats = await this.prismaService.chat.findMany({
        // include: { contact: true },
      })
      if (allChats) {
        return allChats
      } 
      else {
        throw new BadRequestException("chats not found")
      }
    } 
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async findOne(id: string) {
    try {
      const chat = await this.prismaService.chat.findMany({
        where: { id: id },
        // include: { contact: true },
      })
      if (chat) {
        return chat
      } 
      else {
        throw new BadRequestException("chat not found");
      }
    } 
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async getChatByContactId(contactId: string) {
    try {
      const chat = await this.prismaService.chat.findMany({
        where: { contactId: contactId },
        // include: { contact: true },
      })
      if (chat) {
        return chat
      } 
      else {
        console.log("contact not found")
        throw new BadRequestException("contact not found");
      }
    } catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async getChatByUserId(creatorId: string) {
    try {
      const chat = await this.prismaService.chat.findMany({
        where: { creatorId: creatorId },
        // include: { contact: true },
      })
      if (chat) {
        return chat
      } 
      else {
        throw new BadRequestException("chat not found")
      }
    } 
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async update(id: string, data: any) {
    try {
      const result = await this.prismaService.chat.update({
        where: { id: id },
        data: data,
      })
      if (result) {
        return result
      }
      else {
        throw new BadRequestException("error updating chat")
      }
    } 
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async remove(id: string) {
    try {
      const result = await this.prismaService.chat.delete({
        where: { id: id },
      })
      if (result) {
        return { message: "Deleted Successfully" }
      } 
      else {
        throw new BadRequestException("error deleting chat")
      }
    } 
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }

  async searchMessages(text: string) {
    try {
      const messages = await this.prismaService.chat.findMany({
        where: { 
          message: {
            contains: text
          } 
        }
      })
  
      if (messages.length === 0 || messages.length > 0) {
        return messages
      }
      else {
        console.log("No messages found")
        return { message: 'No messages found' }
      }
    }
    catch (error) {
      console.error(error)
      throw new Error("An error occured. Please try again.")
    }
  }
}
