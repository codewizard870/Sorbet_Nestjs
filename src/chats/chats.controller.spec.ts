import { Test, TestingModule } from "@nestjs/testing";
import { ChatsController } from "./chats.controller";
import { ChatsService } from "./chats.service";
import { PrismaService } from "src/utils/prisma/prisma.service";
import { Context, MockContext, createMockContext } from "../../test/prisma/context"
import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

interface CreateChat {
  message: string,
  creatorId: string,
  contactId: string
}

describe("ChatsController", () => {
  let controller: ChatsController;
  let mockCtx: MockContext
  let ctx: Context

  let createChatDto = {
    message: 'hey', 
    contactId: '000102030405060708090a0b'
  }

  let req = {
    user: {
      id: '000102030405060708090a0b'
    }
  }

  const chatData = {
    message: 'hey',
    creatorId: '000102030405060708090a0b',
    contactId: '000102030405060708090a0b'
  }
  const userId = '000102030405060708090a0b'

  let mockChatsService = {
    create: jest.fn().mockImplementation(async (chat: CreateChat) => {
      try {
        const result = await prisma.chat.create({
          data: {
            message: chat.message,
            creatorId: userId,
            contactId: chat.contactId,
          },
        });
        if (result) {
          return result;
        } else {
          throw new Error("result not found");
        }
      } 
      catch (error) {
        console.log(error)
      }
    }),

    getChatByContactId: jest.fn().mockImplementation(async (id: string) => {
      try {
        const chat = await prisma.chat.findMany({
          where: { contactId: id },
        include: { contact: true },
        });
        if (chat) {
          return chat;
        } else {
          throw new Error("contact not found");
        }
      } 
      catch (error) {
        console.log(error)
      }
    }),

    getChatByUserId: jest.fn().mockImplementation(async (id: string) => {
      try {
        const chat = await prisma.chat.findMany({
          where: { creatorId: id },
          include: { contact: true },
        });
        if (chat) {
          return chat;
        } else {
          throw new Error("chat not found");
        }
      } catch (error) {
        console.log(`Error Occured, ${error}`);
      }
    }),

    findOne: jest.fn().mockImplementation(async (id: string) => {
      try {
        const chat = await prisma.chat.findMany({
          where: { id: id },
          include: { contact: true },
        });
        if (chat) {
          return chat;
        } else {
          throw new Error("chat not found");
        }
      } catch (error) {
        console.log(`Error Occured, ${error}`);
      }
    }),

    getAll: jest.fn().mockImplementation(async (id: string) => {
      try {
        const chat = await prisma.chat.findMany({
          include: { contact: true },
        });
        if (chat) {
          return chat;
        } else {
          throw new Error("chat not found");
        }
      } catch (error) {
        console.log(`Error Occured, ${error}`);
      }
    }),

    remove: jest.fn().mockImplementation(async (id: string) => {
      const result = await prisma.chat.delete({
        where: { id: id },
      });
      if (result) {
        return { message: "Deleted Successfully" };
      } else {
        return { message: "Something went wrong" };
      }
    })
  }

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      controllers: [ChatsController],
      providers: [ChatsService, PrismaService],
    })
      .overrideProvider(ChatsService)
      .useValue(mockChatsService)
      .compile();

    mockCtx = createMockContext()
    ctx = mockCtx as unknown as Context
    controller = module.get<ChatsController>(ChatsController);
  });

  it("should be defined", () => {
    expect(controller).toBeDefined()
  })

  // POST - create
  it("create post request should be defined", async () => {
    expect(controller.create).toBeDefined()
  })

  let createdChat: any
  it("should create a chat and post to the database", async () => {
    createdChat = await controller.create(createChatDto, req)
    expect(createdChat).toEqual({
        id: expect.any(String),
        message: 'hey',
        creatorId: '000102030405060708090a0b',
        createdAt: expect.any(Date),
        updatedAt: expect.any(Date),
        contactId: '000102030405060708090a0b'
      })
  })

  // GET - getAll
  it("getAll get request should be defined", () => {
    expect(controller.findAll).toBeDefined()
  })

  it("should get all of the chats and return them in an array", async () => {
    const allChats = await controller.findAll()
    // expect(controller.findAll).toBeCalled()
  })

  // GET - findOne
  it("findOne get request should be defined", () => {
    expect(controller.findOne).toBeDefined()
  })

  it("get a chat by the id and return it", async () => {
    const chat = await controller.findOne(createdChat.id)
    // expect(controller.findOne).toBeCalled()
    // expect(chat).toEqual({
    //   id: expect.any(String),
    //   message: 'hey',
    //   creatorId: '000102030405060708090a0b',
    //   createdAt: expect.any(Date),
    //   updatedAt: expect.any(Date),
    //   contactId: '000102030405060708090a0b'
    // }) 
  })

  it("findBycontactId get request should be defined", () => {
    expect(controller.findBycontactId).toBeDefined()
  })

  it("finds a chat by the contactId and returns it", async () => {
    const chat = await controller.findBycontactId(createdChat.contactId)
    // expect(controller.findBycontactId).toBeCalled()
    // expect(chat).toEqual({
    //   id: expect.any(String),
    //   message: 'hey',
    //   creatorId: '000102030405060708090a0b',
    //   createdAt: expect.any(Date),
    //   updatedAt: expect.any(Date),
    //   contactId: '000102030405060708090a0b'
    // })
  })

  it("findByuserId get request should be defined", () => {
    expect(controller.findByuserId).toBeDefined()
  })

  it("finds a chat by the userId and returns it", async () => {
    const chat = await controller.findByuserId(createdChat.userId)
    // expect(controller.findByuserId).toBeCalled()
    // expect(chat).toEqual({
    //   id: expect.any(String),
    //   message: 'hey',
    //   creatorId: '000102030405060708090a0b',
    //   createdAt: expect.any(Date),
    //   updatedAt: expect.any(Date),
    //   contactId: '000102030405060708090a0b'
    // })
  })

  it("remove delete request should be defined", () => {
    expect(controller.remove).toBeDefined()
  })

  it("removes a chat from the database by id", async () => {
    const removedChat = await controller.remove(createdChat.id)
    expect(removedChat).toEqual({
      message: "Deleted Successfully"
    })
  })
})
