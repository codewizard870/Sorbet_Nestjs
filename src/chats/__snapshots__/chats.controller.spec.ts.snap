// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatsController should create a chat and post to the database 1`] = `
Object {
  "contactId": "000102030405060708090a0b",
  "createdAt": Any<Date>,
  "creatorId": "000102030405060708090a0b",
  "id": Any<String>,
  "message": "hey",
  "updatedAt": Any<Date>,
}
`;
